Add docs on how to connect to JanusGraph Server #205 [skip ci]

FlorianHockmann opened the pull request:
The new part in the docs simply shows examples for the most frequently used languages Java, Python and .NET/C# on how to connect to JanusGraph Server and points to the lists of available TinkerPop drivers and languages. So, users can also just pick one from there in case they are using a language that is not covered by these docs.

These changes intentionally don't mention the new libraries that we currently work on. I propose that we first just add a part to the docs that explains how to connect to JanusGraph from different languages by using the already available TinkerPop GLVs (this PR). We can then extend this part later when we have first official releases for these libraries with #1276.

A preview of the docs is available [here](https://florianhockmann.github.io/docs.janusgraph.org/0.2.2/) with its main new part [Connecting to JanusGraph](https://florianhockmann.github.io/docs.janusgraph.org/0.2.2/connecting.html).

I left out JavaScript simply because I don't have a JS environment set up and because I never used Gremlin-JavaScript. So, it would be good if someone else could add similar docs for JavaScript, but I think that we should do that with a separate PR later on.

Fixes #205
-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [] Have you written and/or updated unit tests to verify your changes?
- [] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
- [x] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

FlorianHockmann commented:
I know I could merge this now via lazy consensus, but I think it would be better if someone would review this as it adds a completely new part to the docs. Again:
> A preview of the docs is available [here](https://florianhockmann.github.io/docs.janusgraph.org/0.2.2/) with its main new part [Connecting to JanusGraph](https://florianhockmann.github.io/docs.janusgraph.org/0.2.2/connecting.html).

hsaputra commented on docs/basics.adoc:
```diff
@@ -344,10 +344,9 @@ This section is a brief overview of the Gremlin query language. For more informa
 * http://tinkerpop.apache.org/docs/{tinkerpop_version}/reference[Complete Gremlin Manual]: Reference manual for all of the Gremlin steps.
 * http://tinkerpop.apache.org/docs/{tinkerpop_version}/tutorials/the-gremlin-console/[Gremlin Console Tutorial]: Learn how to use the Gremlin Console effectively to traverse and analyze a graph interactively.
 * http://tinkerpop.apache.org/docs/{tinkerpop_version}/recipes/[Gremlin Recipes]: A collection of best practices and common traversal patterns for Gremlin.
-* http://tinkerpop.apache.org/index.html#language-drivers[Gremlin Language Drivers]: Connect to a Gremlin Server with different programming languages, including Go, JavaScript, .NET/C#, PHP, Python, Ruby, Scala, and TypeScript.
```
I think we could just leave the link to TinkerPop references for supported drivers.

The additional documentation you added in this PR on how specifically Gremlin can be used to query JanusGraph should help funnel the information from generic TInkerPop to JanusGraph.

chupman commented on docs/connecting-via-python.adoc:
```diff
@@ -0,0 +1,49 @@
+
+Gremlin traversals can be constructed with Gremlin-Python just like in Gremlin-Java or Gremiln-Groovy. Refer to <<gremlin>> for an introduction to Gremlin and pointers to further resources.
+
+IMPORTANT: Some Gremlin step and predicate names are reserved words in Python. Those are simply postfixed with `_` in Gremlin-Python, e.g., `in()` becomes `in_()`, `not()` becomes `not_()`, and so on.
```
I was trying to find a proper link to the reserved words, but [this](https://github.com/apache/tinkerpop/blob/3.3.4/gremlin-python/glv/generate.groovy#L29) is the closest I could find and what I used as a reference when testing the janusgraph-python driver PR. Not sure if this list is exhaustive, but if it is it would be great to include it since linking to source code is less than ideal.

FlorianHockmann replied:
The one you linked should be the best resource as that's what is used to generate Gremlin-Python. I added the affected names to the note.

chupman commented on docs/connecting-via-dotnet.adoc:
```diff
@@ -0,0 +1,46 @@
+----
+var graph = new Graph();
+var client = new GremlinClient(new GremlinServer("localhost", 8182));
+// The client needs to be disposed on shut down to release resources and close open connections
```
In the python example you actually have a comment saying to use connection.close(). I think it would be nice to have a similar comments for .NET and Java. 

Also you used `close` instead of `dispose` in the other two examples. Not sure if dispose is more semantically correct for .NET, but there wasn't much to comment on so I figured I'd include it.

FlorianHockmann replied:
I'll add explicit references to the methods that need to be called for Java and .NET.

Regarding `Dispose()` in .NET: This method is part of the `IDisposable` interface. It is basically always used in cases where resources are released on shut down (like open connections, file handles, ...). This allows to use `using` in C# which looks like this:
```c#
using (var client = new GremlinClient(new GremlinServer("localhost", 8182)))
{
    // do something with the client
}  // <-- `client.Dispose()` is called automatically here even if an exception was thrown
```
`Close()` would be more in line with Gremlin-Java, but it wouldn't feel natural to .NET developers and it couldn't be used with `using`.

FlorianHockmann commented:
Thanks for your reviews @hsaputra and @chupman. I think that I addressed your comments and while working again on this, I also noticed that the config for Java used Gryo v3 which isn't available here as this PR now targets 0.2. So, I changed that to Gryo v1. I'll change the version back to v3 for master during the merge.

hsaputra commented:
LGTM. Sheep it!

Thanks for the great work, Florian.

chupman commented:
I agree. Sheep it!
